x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/async/AsyncTaskIndexService.java

```diff
@@ -187,15 +192,17 @@ public void createResponse(String docId,
                                R response,
                                ActionListener<IndexResponse> listener) throws IOException {
         try {
-            final ReleasableBytesStreamOutput buffer = new ReleasableBytesStreamOutput(0, bigArrays.withCircuitBreaking());
+            final ReleasableBytesStreamOutput buffer = new ReleasableBytesStreamOutputWithLimit(
+                0, bigArrays.withCircuitBreaking(), maxResponseSize);
             final XContentBuilder source = XContentFactory.jsonBuilder(buffer);
-            listener = ActionListener.runBefore(listener, source::close);
+            listener = ActionListener.runBefore(listener, buffer::close);
```

Comment:
Before we were closing `XContentBuilder`, but closing `XContentBuilder` may result in errors:
- if the final `endObject` statement was not present, the `XContentBuilder` will not be closed and will cause a suppressed error: "Failed to close the XContentBuilder"
- another error happens when `XContentBuilder` while closing trying to flush to an underlying buffer, but buffer because of limit or circuit breaker doesn't accept any new input.
 
So this makes sure that the underlying buffer is closed.  I am not sure how it is important to ensure that XContentBuilder is closed as well.

Reply:
I think it's fine to close the buffer instead of XContentBuilder.